Style Loaders & Webpack 3 & babel-preset-env

prescottprue opened the pull request:
* `react` v16.0.0 added to peer deps list
* [`babel-preset-env`](https://github.com/babel/babel-preset-env) used in place of [`babel-preset-es-2015`](https://www.npmjs.com/package/babel-preset-es2015) and [`babel-preset-stage-0`](https://www.npmjs.com/package/babel-preset-stage-0) (fixes npm warning about deprecation)
* `unpkg` parameter added so that UMD version can easily be used in the browser through [the unpkg CDN](https://unpkg.com)
* `react-` prefix removed from name field in `package.json` (unclear when choosing name)
* `css-loader`/`style-loader` config added so that CSS is built in library/module instead of in project that requires the library/module
* npm scripts added:
  * `npm run lint:fix` - check lint and clean up lint where possible
  * `npm run watch` - build unminified version as changes are made (useful when using `npm link`)
  * `npm run clean` - clean build folders (called as first part of build process)
* Dependency updates Including:
  * `react`, `react-dom` and associated - 15.6.2
  * `webpack` - 3.6.0
  * `webpack-dev-server` - 2.8.2
  * Many more

Commit:
Multiple module types (es, commonjs, and umd) as well as other build updates.

* Webpack 3
* babel-preset-env used in place of es-2015 and stage-0
* lint:fix npm script added (`npm run lint:fix`)
* watch npm script added (`npm run watch`)
* clean npm script added (`npm run clean` used in new build command)

Commit:
Working versions of dependency updates (including eslint).

Commit:
No longer exporting commonjs or es version (styles can be affected by project webpack config). CSS loader used in webpack config (correctly exports styling).

zakangelle commented:
Hey man, this looks really good. I'm gonna look at it more in-depth when I get some time tonight. Really appreciate the contribution!

prescottprue commented:
@zakangelle Happy to help 😄 . Thanks for responding so quickly.

I know it is a lot, but let me know if you want explanation/reasoning for on any of it.

Quick note: Main reason for adding the loader config is so that the library/module built with the generator handles building the CSS instead of the project that requires it needing that config.

zakangelle commented:
Really digging these changes.

> react- prefix removed from name field in package.json (unclear when choosing name)

Agreed - not the package's job to enforce package prefixing.

> css-loader/style-loader config added so that CSS is built in library/module instead of in project that requires the library/module

Completely agree.

> npm run watch - build commonJS version as changes are made (useful when using npm link)

Just curious - what is the motivation behind building a CommonJS version? Is this so you can `npm link` multiple adjacent projects and iterate on them in tandem?

Also, unpkg looks really nice.

zxqx commented on app/templates/webpack.config.js:
```diff
@@ -1,14 +1,22 @@
+const isProduction = process.env.NODE_ENV === 'production';
+
+const UglifyJsPlugin = webpack.optimize.UglifyJsPlugin;
+const extractStyles = new ExtractTextPlugin({
```
Looks like this file is missing an import for `ExtractTextPlugin`.

prescottprue replied:
Fixed

zxqx commented on app/templates/webpack.config.js:
```diff
@@ -1,14 +1,22 @@
   output: {
-    path: path.join(__dirname, 'lib'),
-    filename: libraryName + '.min.js',
+    path: path.join(__dirname, 'dist'),
```
Should this be `lib` instead of `dist`? Also, we can prob get rid of any references to `dist` in the project - I was using it as a way to test a prod build of the example app, but it's not really being used anymore.

prescottprue replied:
Yup, I believe you are correct. I believe this is from when I was trying out commonjs/es builds (had an issue getting styling working though).

prescottprue commented:
The thinking behind the commonjs version was to minimize build size since UMDs can often be overkill. Ran into an issue where styles didn't work as expected though, so decided to just leave it at the UMD.

zakangelle commented:
Agreed on UMD being overkill at times - I'm definitely open to ideas.

Commit:
Fixes made based on comments.

webpack.config.js fixes:
  * ExtractTextPlugin is required
  * Build folder switched to lib over dist
  * extractStyles plugin added to plugins list

Commit:
Unamed build goes to lib (unminified version used when using npm link and npm run watch)

prescottprue commented:
Fixed stuff based on your comments and also added something else:

Unnamed version is now built to `lib/index.js` as well as the named versions still being written to `dist/lib-name.js` and `dist/lib-name.min.js`. The `main` field points to the one in `lib`.

This is so that when developing, one can use `npm link` along with `npm run watch` to build unminfied version. Then when running `npm run build`, the minified version is once again built to `lib/index.js` so that this is what can be released.

Commit:
Removed style-loader and added back es and commonjs versions

Renamed FOLDER env var to BUILD_DIR

Commit:
Fix watch command to use commonjs version.

Commit:
Add es to gitignore list.

prescottprue commented:
It seemed like it was easier to just remove the style loader stuff all together. It seems like it is pretty common for component libraries to stick to styling with JS so that CSS files do not have to be handled with loaders ([here is and example of where material-ui](https://github.com/callemall/material-ui/blob/master/src/AppBar/AppBar.js#L9) does it).

With that in mind, I was able to bring back the stuff for making the multiple module types (commonJS, ES2015, and UMD). The watch command builds the commonJS version, but if you are npm linking with a setup that points to the `module` field instead (i.e. webpack 2 or 3). If that is the case it may make more sense to have the watch command build the `es` version instead.

zakangelle commented:
This looks good! Going to go ahead and merge. I have a couple more small things I want to add this weekend, and I'll plan to do a release after.

At some point I do want to revisit how styles are handled since styling with CSS should be an option IMO.

prescottprue commented:
@zakangelle Awesome, thanks. Yeah I totally agree with CSS being an option, just have yet to find a great solution that prevents the project from also need a CSS/style loader setup other than going full JS (i.e. no CSS).

zakangelle commented:
@prescottprue v4.0.0 released - thanks again!
